Tp cryptologie (#50)
prologin/gcc-resources: 08_crypto/08_cryptologie.md

|title|date|
|---|---|
|TP Cryptologie|2021|

|authors|
|---|
|Clarisse 'Nyota' BLANCO|

> |Dorian 'Renji' PÉRON|

Reviewer: Je préfère les mettre dans le md, ils ne s'affichent pas sur le pdf final dans tous les cas je crois.

> |title|date|
> |---|---|
> |TP Cryptologie|2021|
> 
> |authors|
> |---|
> |Clarisse 'Nyota' BLANCO|
> |Dorian 'Renji' PÉRON|

Reviewer 2: ```suggestion
title: Cryptologie
date: 2021
authors: Clarisse Blanco & Dorian Péron
```

# **Le chiffre de César**

_Prérequis : la table ASCII, modulo_

## Un peu d'histoire

---
Ce moyen de chiffrement doit son nom à Jules César, célèbre homme d'état romain. Il aurait utilisé pour certaines de ses communications secrètes, militaires notamment, un chiffrement par substitution. Celui-ci utilisant un décalage d'une valeur trois.

## Qu'est-ce qu'un chiffrement par substitution ?

---
Pour faire simple, cette méthode consiste à substituer la valeur d'une lettre par une lettre, un chiffre ou autre. Dans le cas du chiffrement de César, on substitue une lettre par une autre lettre. Par exemple, si l'on décide d'effectuer un décalage de trois, voici comment la valeur de chaque lettre va évoluer :

![Image not loading](imgs/shifting.jpg)

Voilà comment cela va se répercuter sur l'alphabet tout entier :

| |Substitution avec un décalage de 3|
|---|---|
|Alphabet|A B C D E F G H I J K L M N O P Q R S T U V W X Y Z|
|Substitution|D E F G H I J K L M N O P Q R S T U V W X Y Z A B C|

> On observe donc que A (première lettre de l'alphabet) sera encodée par D (troisième lettre de l'alphabet), que B sera encodée par E, que C par F, etc. On remarque que X sera codé par A, Y par B et Z par C. Si le chiffre donné pour le décalage fait dépasser de l'alphabet, on revient au début.

Reviewer 2: *remplacée

### 2. Le réflecteur

Après avoir fait passer la lettre dans les 3 rotors, on la passe dans une pièce appelée le *réflecteur*. Le réflecteur est une partie similaire à un rotor, car il associe une lettre à une autre, mais reste fixe sans tourner. Enfin, on fini par renvoyer la lettre dans les rotors en sens inverse, et le tour est joué !

### 3. Le tableau de permutations

Comme si cela n'était pas déjà assez compliqué, on rajoute une couche avec le **tableau de permutations**. Le fonctionnement de celui-ci est très simple. Il s'agit simplement de brancher des lettres 2 par 2 pour échanger leur place (on peut échanger comme cela jusqu'à 10 paires de lettres). On branche ce tableau avant le système de rotor. Par exemple, si l'on échange les lettres **A** et **O**, alors chaque **A** tapé par l'opérateur de la machine deviendra un **O** pour le système des rotors, et chaque **A** "renvoyé" par ce dernier sera un **O** pour l'opérateur (et vice-versa, le **O** devient un **A**).

> Schéma de fonctionnement d'Enigma
> 

Reviewer 3: ```suggestion
```

> !["schéma fonctionnement"](imgs/enigma_scheme.jpg "Schéma de fonctionnement")

Reviewer 3: Les numéros des rotors sont inversés par rapport aux explications non ?

Reviewer 2: c'est possible, en effet. mais en même temps je me dis que "le signal passe d'abord dans le 1er rotor" c'est plus naturel que "le signal passe d'abord dans le 3e rotor"

> !["schéma fonctionnement"](imgs/enigma_scheme.jpg "Schéma de fonctionnement")

Reviewer 3: ```suggestion
![Schéma de fonctionnement d'Enigma](imgs/enigma_scheme.jpg)
```

> >Si ces explications manquent de clarté, voici [un site très pratique pour visualiser le chemin d'une lettre dans le système](https://observablehq.com/@tmcw/enigma-machine)

Reviewer 3: ```suggestion
Vous pouvez aussi visualiser le fonctionnement d'Enigma sur ce site : https://observablehq.com/@tmcw/enigma-machine
```

### 4. Et pour décoder un message ?

> Tu viens d'apprendre le fonctionnement général de la machine Enigma, mais une question persiste : Avec un système aussi complexe, comment fait-on pour décoder un message crypté avec une machine Enigma ? La réponse est sûrement plus simple qu'il n'y paraît : Il suffit de taper le message chiffré sur une machine Enigma qui a les mêmes paramètres initiaux que pour encoder le message ! En effet, prenons le message "GirlsCanCode" et imaginons que tous les rotors soient sur le cran 0 au début du message. Taper **G** sur la machine donnera une lettre différente, peut-être un **V**. Mais de la même manière, si l'on avait tapé **V**, on aurait obtenu **G**, car le chemin de la lettre dans la machine est le même, seulement il est parcouru à l'envers.

Reviewer 3: ```suggestion
Tu viens d'apprendre le fonctionnement général de la machine Enigma, mais une question persiste : avec un système aussi complexe, comment fait-on pour déchiffrer un message chiffré avec une machine Enigma ? La réponse est sûrement plus simple qu'il n'y paraît : il suffit de taper le message chiffré sur une machine Enigma qui a les mêmes paramètres initiaux que pour chiffrer le message ! En effet, prenons le message "GirlsCanCode" et imaginons que tous les rotors soient sur le cran 0 au début du message. Taper **G** sur la machine donnera une lettre différente, peut-être un **V**. Mais de la même manière, si l'on avait tapé **V**, on aurait obtenu **G**, car le chemin de la lettre dans la machine est le même, seulement il est parcouru à l'envers.
```

Reviewer 3: Je le met ici mais c'est valable pour un peu partout :

En français :
:x: crypter
:heavy_check_mark: chiffrer
:x: coder, encoder, décoder
:heavy_check_mark: chiffrer, déchiffrer

:heavy_check_mark: décrypter (quand on ne connait pas la clé)

En anglais :
:heavy_check_mark: encrypt, decrypt (je ne crois pas qu'il y ait de distinction entre déchiffrer et décrypter)
:grey_question: cipher, decipher (je ne sais pas si c'est vraiment correct / utilisé)

## **Exercices**

À ton tour de créer ton propre programme Enigma !

> > Afin d'implémenter notre propre machine Enigma, nous utiliserons une classe qui nous sera utile pour sauvegarder l'état actuel des rotors.

Reviewer 3: ```suggestion
Afin d'implémenter notre propre machine Enigma, nous utiliserons une classe qui nous sera utile pour sauvegarder l'état actuel des rotors.
```

Voici le squelette du code:

```py
class Enigma:
    def __init__(self, r1, r2, r3, ref):
        """
        Initialise la machine avec les rotors et le réflecteur donnés
        Appelé par Enigma(...)
        r1 : rotor1
        r2 : rotor2
        r3 : rotor3
        ref : reflecteur
        """
        self.r1 = r1
        self.r2 = r2
        self.r3 = r3
        self.ref = ref
        
        # on initialise le décalage des rotors à 0
        self.x1 = 0
        self.x2 = 0
        self.x3 = 0
    
    def reset_rotors(self):
        # on réinitialise le décalage des rotors à 0
        self.x1 = 0
        self.x2 = 0
        self.x3 = 0
    
    def encode_letter(self, letter):
        pass # À toi de jouer !

    
    def encode_message(self, msg):
        pass # À toi de jouer !

# Exemples

# on déclare notre variable Enigma (avec les rotors définis ci-dessous)

> # état initial: tous les décalages à 0

Reviewer 3: ```suggestion
# état initial : tous les décalages à 0
```

>>> en = Enigma(rotor_I, rotor_II, rotor_III, reflector_A)

>>> en.encode('A')
'N' 

> # Etat interne des rotors après opération: x1 = 1, x2 = 0, x3 = 0

Reviewer 3: ```suggestion
# État interne des rotors après opération: x1 = 1, x2 = 0, x3 = 0
```

```

> Voici également 2 simples fonctions permettant de trouver le nombre d'une lettre dans l'alphabet (en partant de `0`), et vice-versa:

Reviewer 3: ```suggestion
Voici également 2 simples fonctions permettant de trouver l'indice d'une lettre dans l'alphabet (en partant de `0`), et vice-versa:
```

À voir si on ne leur donne pas plus tôt ? Ça sert aussi pour César, etc.

```py
def letter_to_number(letter):
    """
    Donne la position d'une lettre dans l'alphabet
    """
    letter = letter.upper() # on transforme la lettre en majuscule
    return ord(letter) - ord('A') 

def number_to_letter(number):
    """
    Donne la lettre majuscule correspondante au nombre donné
    """
    return chr(number + ord('A'))

> # Examples

Reviewer 3: ```suggestion
# Exemples
```

>>> letter_to_number('A')
0
>>> letter_to_number('z')
25

>>> number_to_letter(0)
A
>>> number_to_letter(23)
X
```

> Enfin, voici des rotors et réflecteurs qui ont réellement été utilisés pendant la seconde guerre mondiale:

Reviewer 3: ```suggestion
Enfin, voici des rotors et réflecteurs qui ont réellement été utilisés pendant la seconde guerre mondiale :
```

```py
rotor_I = "EKMFLGDQVZNTOWYHXUSPAIBRCJ"
rotor_II = "AJDKSIRUXBLHWTMCQGZNPYFVOE"
rotor_III = "BDFHJLCPRTXVZNYEIWGAKMUSQO"

reflector_A = "EJMZALYXVBWFCRQUONTSPIKHGD"

> reflector_B = "YRUHQSLDPXNGOKMIEBFZCWVJAT"

Reviewer 3: Pourquoi il y deux réflecteurs ?

Reviewer 2: au cas où, pour changer si on veut

```

---

### **Partie 1 : encode_letter()**

Dans cette méthode de la classe Enigma, tu dois crypter la lettre donnée en paramètre.

> >Entrée(s):
> 
> - letter: la lettre à chiffrer
> 
> >Sortie:
> 
> - La lettre chiffrée correspondante

Reviewer 3: ```suggestion
Entrée :
- letter: la lettre à chiffrer

Sortie :
- La lettre chiffrée correspondante
```

**ATTENTION**: après avoir trouvé la lettre, il ne faut pas oublier de **modifier la position des rotors**.

---

### **Partie 2 : encode_message()**

Dans cette méthode, il te faut encoder tout un message avec le programme Enigma !

> >Entrée(s):
> 
> - msg: le message à chiffrer
> 
> >Sortie: Le message chiffré
> 
> - N'oublie pas que tu peux utiliser la méthode `encode_letter()` dans `encode_message()`.
> - Attention aux caractères spéciaux dans ton message, les chiffres, espaces et caractères de ponctuation ne peuvent pas être chiffrés par Enigma.

Reviewer 3: ```suggestion
Entrée :
- msg: le message à chiffrer

Sortie :
- Le message chiffré

- N'oublie pas que tu peux utiliser la méthode `encode_letter()` dans `encode_message()`.
- Attention aux caractères spéciaux dans ton message, les chiffres, espaces et caractères de ponctuation ne peuvent pas être chiffrés par Enigma.
```

> Une fois implémentée, tu peux tester ton code de la manière suivante:

Reviewer 3: ```suggestion
Une fois implémentée, tu peux tester ton code de la manière suivante :
```

```py
>>> msg = 'HELLO WORLD'
>>> en.reset_rotors()
>>> coded = en.encode_message(msg)
>>> print(coded)
# message chiffré
>>> en.reset_rotors()
>>> decoded = en.encode_message(coded)
>>> print(decoded)
# 'HELLOWORLD'
```

> Si tous fonctionne normalement, le programme devrait afficher la version chiffrée puis déchiffrée de ton message !

Reviewer 3: ```suggestion
Si tout fonctionne normalement, le programme devrait afficher la version chiffrée puis déchiffrée de ton message !
```

---

### **Partie 3 : Bonus**

> Bien joué!

Reviewer 3: ```suggestion
Bien joué !
```

> Pour aller plus loin:
> 
> > Tu peux modifier le programme pour ajouter la fonctionnalité du tableau de permutation, qui n'est pas montrée ici.

Reviewer 3: ```suggestion
Pour aller plus loin, tu peux modifier le programme pour ajouter la fonctionnalité du tableau de permutation, qui n'est pas montrée ici.
```

> > **Info**: une particularité de la machine Enigma qui a aidé A. Turing et son équipe de casser le code est la suivante : Enigma ne donnait jamais de lettre identique à celle donnée en entrée. Mais peux-tu expliquer pourquoi ?

Reviewer 3: ```suggestion
> **Info**: une particularité de la machine Enigma qui a aidé Alan Turing et son équipe pour casser le code est la suivante : Enigma ne donnait jamais de lettre identique à celle donnée en entrée. Saurais-tu expliquer pourquoi ?
```

# **XOR / Ou Exclusif**

> >Prérequis : l'hexadécimal, la table ASCII

Reviewer 3: ```suggestion
_Prérequis : l'hexadécimal, la table ASCII_
```

## Quelques rappels

---

> _Explication du binaire à ajouter ?_

Reviewer 3: Étant donné qu'on donne un exemple en binaire plus bas, ça me semble bien d'en parler un peu

### **L'hexadécimal**

> L'hexadécimal est un système de comptage en base 16. Il utilise 16 symboles : les lettres de A à F (inclus) et les chiffres de 0 à 9 (inclus).

Reviewer 3: ```suggestion
L'hexadécimal est un système permettant d'écrire des nombres en utilisant 16 symboles : les chiffres de 0 à 9 (inclus), puis les lettres de A à F (inclus).
```

À titre d'exemple, voici comment compter de 0 à 20 en hexadécimal :

```sh
base 10 : 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
base 16 : 0 1 2 3 4 5 6 7 8 9  A  B  C  D  E  F 10 11 12 13 14
```

> **Remarque :** on utilise le système décimal, en base 10, dans la vie courante. Il utilise donc 10 symboles : les chiffres de 0 à 9.

### **La table ASCII**

> La table ascii (American Standard Code for Information Interchange), est une norme d'encodage des caractères. C'est une représentation qu'utilisent les ordinateurs pour garder en mémoire des caractères. Par exemple, avec la table ASCII, le `A` correspond à `65` et le `a` correspond à `97`.

Reviewer 3: ```suggestion
La table ASCII (American Standard Code for Information Interchange), est une norme d'encodage des caractères. C'est une représentation qu'utilisent les ordinateurs pour garder en mémoire des caractères. Par exemple, avec la table ASCII, le `A` correspond à `65` et le `a` correspond à `97`.
```

Là "encodage" c'est good :ok_hand:

> ![Ascii](../imgs/ascii.png "Table ASCII")

Reviewer 3: ```suggestion
![Table ASCII](../imgs/ascii.png)
```

## Un peu de théorie

---

> Le XOR est un opérateur logique, on l'appelle également le **OU Exclusif** et en anglais **eXclusive OR**. Il est souvent représenté avec le symbole :  **⊕** ". En Python, on utilisera : **^**.

Reviewer 3: Le symbole n'est pas correctement affiché en LaTeX

Table de vérité :

- On définit A et B deux opérandes binaires sur lesquelles on va appliquer XOR.

|A|B|A ⊕ B|
|---|---|---|
|0|0|0|
|0|1|1|
|1|0|1|
|1|1|0|

> En observant la table de vérité, on peut dire que le XOR peut se définir de la façon suivante:

Reviewer 3: ```suggestion
En observant la table de vérité, on peut dire que le XOR peut se définir de la façon suivante :
```

- Si les deux opérandes sont différentes, alors le résultat est 1
- Si les deux opérandes sont identiques, alors le résultat est 0

## Chiffrer des messages avec XOR

Dans cet exercice, vous allez chiffrer des messages avec XOR. Cet opérateur logique possède en effet des propriétés intéressantes qui vont permettre de chiffrer et déchiffrer un message avec la même clé.

Pour mieux vous faire comprendre le processus, appliquons-le avec un exemple en binaire. Admettons que le message que l'on souhaite chiffrer est `01100101 01110000 01101001` et que la clé est `01110100 01100001 01100001`.

> ```sh

Reviewer 3: ```suggestion
```text
```

Maintenant que vous connaissez le XOR, il est temps d'écrire les fonctions !

## Let's code

---
Pour implémenter cette méthode de chiffrement, il va être plus simple de diviser le problème en plusieurs parties, et donc en plusieurs fonctions. Comme écrit précedemment, pour appliquer le XOR, pas besoin de passer par du binaire, nous allons utiliser de l'hexadécimal. La première fonction va donc transformer du texte en hexadécimal. La deuxième va appliquer XOR sur le message transformé en hexadecimal avec une clé en hexadecimal de la même longueur. Finalement, la troisième va transformer de l'hexadecimal en lettres, texte lisible pour nous !

**Comment obtenir une clé en hexadécimal de la même longueur que le message que l'on souhaite chiffrer ?**

C'est tout simple, une fonction générant des clés en hexadécimal vous est fournie : `randomHexa(lenstr)`. Elle prend en paramètre `lenstr` qui correspond à la longueur du message initial en lettres et renvoie une clé générée aléatoirement. Cette fonction vous sera grandement utile pour effectuer vos tests !

Voici la fonction :

```py

> import random

Reviewer 3: ```suggestion
import random  # n'oubliez pas cette ligne !
```

def randomHexa(lenstr):
    string = ""
    for i in range(lenstr):

>         rdm_int = random.randint(0, 255) # on génère un entier entre 0 et 255
>         string += chr(rdm_int) # on convertit en caractère et on concatène
>     res = stringToHexa(string) # on convertit en hexa pour avoir une clé en hexa

Reviewer 3: ```suggestion
        rdm_int = random.randint(0, 255)  # on génère un entier entre 0 et 255
        string += chr(rdm_int)  # on convertit en caractère et on concatène
    # Attention, vous devez coder la fonction `stringToHexa`, sinon ça ne fonctionnera pas
    res = stringToHexa(string)  # on convertit en hexa pour avoir une clé en hexa
```

J'avoue là j'abuse un peu, mais la PEP8 nous dit de mettre 2 espaces avant un commentaire :wink:

:warning: À appliquer partout où c'est nécessaire

>>> hex(16)
'0x10'
>>> hex(42)
'0x2a'

# Pour retirer un nombre x de lettres au début d'un mot, on utilise [x:]
>>> s = "moustique"
>>> s[4:]
'tique'

#La fonction zfill() prend en paramètre un nombre entier. 
#Elle permet de rajouter des 0 au début d'une chaîne de caractères 
#si la longueur de celle-ci est inférieure au paramètre donné à la fonction. 
>>>'hello'.zfill(6)
'0hello'
>>> 'hello'.zfill(5) # Si le paramètre est inférieur ou égal à la taille de la chaîne de caractère, il n'y a pas de changement
'hello'
```

### La fonction `msgXor(m, key)`

Cette fonction va se charger d'appliquer XOR entre le message et la clé (soit message ⊕ clé). Elle prend donc en paramètre `m` une chaîne de caractères correspondant au message que l'on désire chiffrer, et **qui a déjà été converti en hexadecimal**. Elle prend également `key` qui est également une chaîne de caractères en hexadecimal et qui est de la même longueur que `m`. `msgXor(m, key)` va renvoyer une chaîne de caractères en hexadecimal.

> Pour appliquer XOR, voici les différentes étapes qu'il faudra implémenter:

Reviewer 3: ```suggestion
Pour appliquer XOR, voici les différentes étapes qu'il faudra implémenter :
```

- Les convertir en nombre entier
- Appliquer XOR !

**Quelques fonctions utiles...**

```py
# La fonction int() va convertir un caractère ou une chaîne de caractère en un entier. Dans notre cas, nous allons renseigner deux paramètres : une chaîne de caractères en hexadécimal et la base de notre chaîne de caractères (ici, 16 car on utilise de l'hexadecimal). 
>>> int('10', 16)
 16
>>> int('FA', 16)
250

# L'opérateur logique XOR est symbolisé par ^ en Python. Il peut être utilisé avec des nombres entiers.
>>> 42^56
18

# Les fonctions et opérateurs hex(), [x:], zfill(), expliqués plus haut.
```

### La fonction `hexaToString(hexa)`

Cette fonction va se charger de convertir une chaîne de caractères de valeurs hexadécimales, en une chaîne de caractères de lettres, que l'on pourra lire et comprendre. Elle va donc renvoyer une chaîne de caractères. `hexaToString(hexa)` prend en paramètre `hexa`, une chaîne de caractères en hexadécimal.

Les fonctions dont vous aurez besoin pour implémenter votre solution ont déjà été expliquées plus haut. Les étapes ne sont pas données mais vous pouvez les déduire des étapes de `stringToHexa()`.

À vous de jouer !

## Sources

---

> - [Table ASCII](https://computersciencewiki.org/images/3/3d/Ascii_table.png)
> 
> - [Infos sur la table ASCII](http://www.table-ascii.com/)

Reviewer 3: ```suggestion
- [Table ASCII](https://computersciencewiki.org/images/3/3d/Ascii_table.png)
- [Infos sur la table ASCII](http://www.table-ascii.com/)
```
